databroker/src/broker.rs

```diff
@@ -1561,6 +1623,290 @@ impl<'a, 'b> AuthorizedAccess<'a, 'b> {
+                    return Err((
+                        ActuationError::PermissionExpired,
+                        "Permission expired".to_string(),
+                    ))
+                }
+            }
+        }
+
+        let provided_vss_ids: Vec<i32> = self
+            .broker
+            .subscriptions
+            .read()
+            .await
+            .actuation_subscriptions
+            .iter()
+            .flat_map(|subscription| subscription.vss_ids.clone())
+            .collect();
+        let intersection: Vec<_> = vss_ids
+            .iter()
+            .filter(|&x| provided_vss_ids.contains(x))
+            .collect();
+        if !intersection.is_empty() {
+            let message = format!(
+                "actuation providers for the following vss_ids already registered: {:?}",
+                intersection
+            );
+            return Err((ActuationError::ProviderAlreadyExists, message));
+        }
+
+        let actuation_subscription: ActuationSubscription = ActuationSubscription {
+            vss_ids,
+            actuation_provider,
+            permissions: self.permissions.clone(),
+        };
+        self.broker
+            .subscriptions
+            .write()
+            .await
+            .add_actuation_subscription(actuation_subscription);
+
+        Ok(())
+    }
+
+    async fn map_actuation_changes_by_vss_id(
+        &self,
+        actuation_changes: Vec<ActuationChange>,
+    ) -> HashMap<i32, Vec<ActuationChange>> {
+        let mut actuation_changes_per_vss_id: HashMap<i32, Vec<ActuationChange>> = HashMap::new();
+        for ele in actuation_changes {
+            let vss_id = ele.id;
+
+            let opt_vss_ids = actuation_changes_per_vss_id.get_mut(&vss_id);
+            match opt_vss_ids {
+                Some(vss_ids) => {
+                    vss_ids.push(ele.clone());
+                }
+                None => {
+                    let vec = vec![ele.clone()];
+                    actuation_changes_per_vss_id.insert(vss_id, vec);
+                }
+            }
+        }
+        actuation_changes_per_vss_id
+    }
+
+    pub async fn batch_actuate(
+        &self,
+        actuation_changes: Vec<ActuationChange>,
+    ) -> Result<(), (ActuationError, String)> {
+        let actuation_subscriptions = &self
+            .broker
+            .subscriptions
+            .read()
+            .await
+            .actuation_subscriptions;
+
+        for actuation_change in &actuation_changes {
+            let vss_id = actuation_change.id;
+            let result_entry = self.get_entry_by_id(vss_id).await;
+            match result_entry {
+                Ok(entry) => {
+                    let vss_path = entry.metadata.path;
+                    let result_can_write_actuator =
+                        self.permissions.can_write_actuator_target(&vss_path);
+                    match result_can_write_actuator {
+                        Ok(_) => {}
```

Comment:
Shouldn't we check here if a provider is available else we can break here no need to do two for loops. You can fill the map<vss_id, actuation_change> here as well. And then you just emit the actuation requests once you collected all actuation changes successfully

Reply:
Would be possible. However I tried to not mix to many concerns here. Therefore this loop just cares about the permissions.

Having multiple loops shouldn't be the issue however having to do multiple id lookups might be bad performance-wise. Since the creation of the Map<vss_id, Vec<ActuationChange>> is a complex method on it's own I would prefer leaving it in it's own method instead of inlining the code inside of the ok block. Here I vote for readability / maintainability over performance.

However, if you have a good idea how to achieve both, please feel free to make a short pseudo-code suggestion (or let's have another call).